ksqldb-rest-app/src/test/java/io/confluent/ksql/rest/integration/PauseResumeIntegrationTest.java

```diff
@@ -0,0 +1,279 @@
+
+    TEST_HARNESS.ensureTopics(PAGE_VIEW_TOPIC);
+    RestIntegrationTestUtil.createStream(REST_APP, PAGE_VIEWS_PROVIDER);
+  }
+
+  @After
+  public void cleanUp() {
+    REST_APP.closePersistentQueries();
+    REST_APP.dropSourcesExcept();
+    REST_APP.stop();
+  }
+
+  @Test
+  public void shouldPauseAndResumeQuery() {
+    // Given:
+    createQuery("");
+    TEST_HARNESS.produceRows(PAGE_VIEW_TOPIC, PAGE_VIEWS_PROVIDER, KAFKA, JSON, System::currentTimeMillis);
+
+    Supplier<Integer> supplier = () -> RestIntegrationTestUtil.makeQueryRequest(REST_APP,
+        "select * from " + SINK_STREAM + ";",Optional.empty()).size();
+
+    // 7 records are produced + a header & footer.
+    assertThatEventually(supplier, equalTo(9));
+
+    // When:
+    // JNH: factor out?
+    String queryId = ((Queries) RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "SHOW "
+        + "QUERIES;").get(0)).getQueries().get(0).getId().toString();
+    RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "PAUSE " + queryId + ";");
+
+    assertThat(getPausedCount(), equalTo(1));
+
+    // Produce more records
+    TEST_HARNESS.produceRows(PAGE_VIEW_TOPIC, PAGE_VIEWS_PROVIDER2, KAFKA, JSON,
+        System::currentTimeMillis);
+
+    // Observe no new records -- This kinda *doesn't work*.  Hard to prove a negative sometimes...
+    assertThatEventually(supplier, equalTo(9));
+
+    // Then:
+    RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "RESUME "
+        + queryId + ";");
+
+    // 5 more records have been produced
+    assertThatEventually(supplier, equalTo(14));
+    assertThat(getRunningCount(), equalTo(1));
+  }
+
+  @Test
+  public void shouldPauseAndResumeMultipleQueries() {
+    // Given:
+    createQuery("1");
+    createQuery("2");
+
+    Supplier<Integer> supplier = getSupplier(1);
+    TEST_HARNESS.produceRows(PAGE_VIEW_TOPIC, PAGE_VIEWS_PROVIDER, KAFKA, JSON, System::currentTimeMillis);
+
+    // 7 records are produced + a header & footer.
+    assertThatEventually(supplier, equalTo(9));
+
+    // When:
+    String queryId = ((Queries) RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "SHOW "
+        + "QUERIES;").get(0)).getQueries().get(0).getId().toString();
+    RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "PAUSE " + queryId + ";");
+    assertThat(getPausedCount(), equalTo(1));
+
+    // Produce more records
+    TEST_HARNESS.produceRows(PAGE_VIEW_TOPIC, PAGE_VIEWS_PROVIDER2, KAFKA, JSON,
+        System::currentTimeMillis);
+    
+    assertThatEventually(supplier, equalTo(9));
+
+    // Then:
+    RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "RESUME " + queryId + ";");
+
+    // 5 more records have been produced
+    assertThatEventually(supplier, equalTo(14));
+
+    List <KsqlEntity> showQueries3 = RestIntegrationTestUtil.makeKsqlRequest(REST_APP, "SHOW "
+        + "QUERIES;");
+    assertThat(((Queries) showQueries3.get(0)).getQueries().get(0).getStatusCount().getStatuses()
+        .get(KsqlQueryStatus.RUNNING), equalTo(1));
```

Comment:
Can we add a status check into this test for when the query is paused too?

Reply:
Yes, I'm trying to add a little bit for that.